### ISSUE TITLE
Open Request Cost Aggregation (ORCA)

### ISSUE THREAD

**Issue by htuch** [envoyproxy/envoy#6614]
Today in Envoy, simple load balancing decisions can be made by taking into account local or global knowledge of a backend’s load, for example CPU. More sophisticated load balancing decisions are possible with application specific knowledge, e.g. queue depth, or by combining multiple metrics.

This is useful for services that may be resource constrained along multiple dimensions (e.g. both CPU and memory may become bottlenecks, depending on the applied load and execution environment, it’s not possible to tell which upfront) and where these dimensions do not slot within predefined categories (e.g. the resource may be “number of free threads in a pool”, disk IOPS, etc.).

https://docs.google.com/document/d/1NSnK3346BkBo1JUU3I9I5NYYnaJZQPt8_Z_XCBCI3uA/edit# provides a design proposal for an Open Request Cost Aggregation (ORCA) standard for conveying this information between proxies like Envoy and upstreams. We propose that this become a standard part of UDPA and supported by Envoy.

The design document is in draft stage; from offline discussions I think the need for something like this is not very controversial, we can iterate on aspects of the design here.

**Comment by SecurityInsanity** [comment]
I've been talking with @htuch about implementing ORCA's in-band reporting, and adding it's details to a particular stream.)

However, when bringing it up @htuch mentioned that the RFC had some debate around _how_ inband reporting should be implemented. Currently the RFC Calls for stuffing JSON in `x-endpoint-load-metrics`. However after talking with @PiotrSikora I think this is the incorrect choice, and would like to spur extra conversation here. Since I shouldn't be the sole person to make this decision :stuck_out_tongue_winking_eye:  

---

While the JSON Reporter is using a standard encoding (JSON), not all proxies currently support JSON out of the box, nor should they. Most of the time the bytes are just moving through them, and they only need to know how to parse HTTP. (Some can be extended: see NGINX, but those are always custom extensions/code written to do so). 

If we want ORCA to become a true standard, by lowering the barrier to entry for them by not forcing them to add in JSON parsing if they don't need to, adoption would increase. Which would help the total number of users, and it's usefulness.

Instead I recommend we implement ***both*** parsing of: `x-endpoint-load-metrics-bin` (binary protobuf format), and: `x-endpoint-load-metrics`. The real one we should focus on is: `x-endpoint-load-metrics` (which maybe should even be called `endpoint-load-metrics` since the IETF recommends not using [x-](https://tools.ietf.org/html/rfc6648)). The reason for this is two fold:

   1. It's an RFC that's [very close to completing](https://tools.ietf.org/html/draft-ietf-httpbis-header-structure), and already has other RFCs building [on top of it](https://tools.ietf.org/html/draft-ietf-httpbis-proxy-status-00). So it's chances of "dieing out" are low.
    2. Even headers that are widely supported today are using something very close to parameter list: `Cache-Control: max-age=seconds` for example. This is already the basis for parameter list.

`x-endpoint-load-metrics-bin` I think should be supported because it can be a nice optimization for those already integrating with ORCA protobuf, and wanting to not have to juggle two seperate encodings of what to send on.I don't imagine it being a huge ask to do so, making it not worth the implementation effort.

---

Happy to hear thoughts on this, and coming up with something official that aren't just my bemused thoughts :smile:

**Comment by htuch** [comment]
@SecurityInsanity yes, https://tools.ietf.org/html/draft-ietf-httpbis-header-structure-11 is fine. Let's just make sure that this is a direct translation of the data model that exists in the proto, i.e. it should be equivalently expressive.

**Comment by SecurityInsanity** [comment]
For sure 👍🏻. I think keeping the model is needed. 

I’ll start working on a PR tomorrow. Can you update the doc? (I don’t think I have write access).

**Comment by htuch** [comment]
@SecurityInsanity sure. I think we should hash out the new representation here first. Looking at https://github.com/envoyproxy/envoy/blob/master/api/udpa/data/orca/v1/orca_load_report.proto, I think we may need multiple headers, e.g. `x-endpoints-load-metrics`, `x-endpoints-load-metrics-cost`, `x-endpoint-load-metrics-utilization` to correctly distinguish the core fields and the distinct maps that exist in the data model. @PiotrSikora do you think this is correct?

FWIW, coincidentally I'm working on migrating the API tree in https://github.com/envoyproxy/envoy/tree/master/api/udpa (which includes the ORCA protos) to live in https://github.com/cncf/udpa-wg today. This shouldn't have a major impact on your work, but there might need to be some slight path or Bazel fixups needed once this lands.

**Comment by SecurityInsanity** [comment]
@htuch That's good to know about the proto's moving, thanks. Based on my understanding of the priority yes we'd need 3 headers. (one for the core type, and two for the two maps).

**Comment by htuch** [comment]
@SecurityInsanity sounds like a plan from my side. Looking forwarding to seeing ORCA support landing :)

**Comment by htuch** [comment]
@SecurityInsanity assigning issue to you for the ORCA implementation work planned. Feel free to assign back if there is any remaining future work once that lands.

**Comment by CodingSinger** [comment]
Hello everyone, I have a question, is there any difference between orca_load_report and the original LRS? My understanding is that orca_load_report is the backend server passing load information to envoy, and LRS is passing information between envoy and management server？

**Comment by SecurityInsanity** [comment]
Hey @CodingSinger,

ORCA for now is actually going to be integrated into the LRS when it is implemented. It will provide a richer set of information.

Right now the LRS only provides load info about the number of requests, who it’s routing to and when. ORCA compliments that info by allowing services to report how much a request cost. For example a service can say “processing this request I took up 20% cpu”. 

There are two ways a service can report this back to envoy:

- Through headers in the response.
- Through a separate out of band reporting mechanism.

We’re targeting reading response headers first. Admittedly I’ve had a lot going on so this has slumped, however I hope to have something up in the coming weeks.

**Comment by CodingSinger** [comment]
@SecurityInsanity 
Thanks for your reply. But I think now it seems to be divided into `LoadReportingService` and` OrcaLoadReport`. According to your reply, are both ORCA and LRS acting between the backend server and envoy?

**Comment by SecurityInsanity** [comment]
@CodingSinger ,

ORCA metrics will be added to the LoadReportingService (not replacing) stats since we believe it is useful there as well, but the actual ORCA stats are being sent between the thing envoy is sending requests to and envoy.

**Comment by CodingSinger** [comment]
@SecurityInsanity Thanks.
I have got it.

**Comment by erikbos** [comment]
Does anybody have any pointers to blogs / papers about about considerations for multi region or global load balancing algorithms? (Useful for input into a designing a system which would leverage functionality like ORCA)

**Comment by htuch** [comment]
That's a great question @erikbos. @alexburnos @antoniovicente are you folks aware of any public material that talks to how backend named costs would integrate with global LB?

**Comment by alexburnos** [comment]
Do know anything public that would be specifically focused on LB algorithms, but maybe [chapter on managing load in the SRE book](https://landing.google.com/sre/workbook/chapters/managing-load/) could give some high level ideas.

**Comment by erikbos** [comment]
Thanks for the reference to the SRE book, it's always a good read but I was looking for the next level of depth.. On Slack @snowp mentioned https://netflixtechblog.com/netflix-edge-load-balancing-695308b5548c which contains some of that 👍

**Comment by htuch** [comment]
gRPC has adopted ORCA (and its xDS definitions) as the basis of load reporting for gRPC-LB v2 CC @markdroth. We still do not have any Envoy implementation though, very much open to any contribution PRs here.

**Comment by markdroth** [comment]
Just for reference for anyone working on this, the ORCA support in gRPC is documented in [gRFC A51: Custom Backend Metrics Support](https://github.com/grpc/proposal/blob/master/A51-custom-backend-metrics.md) and [gRFC A64: xDS LRS Custom Metrics Support](https://github.com/grpc/proposal/blob/master/A64-lrs-custom-metrics.md).

See also [gRFC A58: weighted_round_robin LB policy](https://github.com/grpc/proposal/blob/master/A58-client-side-weighted-round-robin-lb-policy.md) for how ORCA is used in load balancing.

**Comment by soulxu** [comment]
@Mythra are you still working on this? If not, I'm a little interesting in this issue.

**Comment by htuch** [comment]
@efimki is working on this from our side (Google). CC @markdroth @AndresGuedez

**Comment by osswangxining** [comment]
> @efimki is working on this from our side (Google). CC @markdroth @AndresGuedez

Any detail info about this? Looking forward to this feature. Any we can help this?

**Comment by efimki** [comment]
Here is a draft outline of what we are trying to do: 

- Load Reports will be provided to the xDS control plane server via xDS LRS API.
- Load reports will be used by a new Client Side Weighted Round Robin load balancing policy to dynamically calculate host weights on the client side. Inline reporting enables sub second load balancing reaction times (depending on backend load measurement and reporting intervals), a critical requirement for customers with coordinated and spiky traffic workloads.
- Using these load reports, Envoy proxies will be able to implement load balancing policies that vary endpoint load balancing weights according to backend load reports.

More details are [here](https://docs.google.com/document/d/1gb_2pcNnEzTgo1EJ6w1Ol7O-EH-O_Ysu5o215N9MTAg/edit?usp=sharing).

**Comment by wbpcode** [comment]
Basically, I think there are two different part works:
1. bridge the orca report and LRS.
2. make the lb aware the orca.

I personally think we should only provide simplest support to the common metrics: cpu, mem, application_utilization, etc. first. There attributes could cover most cases.

The `named_metrics`, `utilization`, and `request_cost` may has more complex semantics and will bring more heavy overhead. So, I will prefer to only provide simplest implementation first until our users ask more.

**Comment by efimki** [comment]
I agree with a two parts distinction.

We will start with using orca report for LRS. I agree that common orca metrics cover many cases, however we want to provide our users with flexibility of using named metrics if necessary. The additional complexity of handling named metrics on top of processing orca report is not that high.

**Comment by 0lut** [comment]
is there any movement on this? i am interested in using this feature

**Comment by wbpcode** [comment]
> is there any movement on this? i am interested in using this feature

It's landed but now only client weighted round robin will use the report to help load balancing.

**Comment by jukie** [comment]
Is there another proposed load balancer this would fit under or is there a desire to consider LocalityLbConfig? I'm interested in contributing towards additional ORCA support but would like to understand the current state and outstanding features.

**Comment by myf5** [comment]
Is ORCA now widely supported by various LLM Inference runtime?

**Comment by holooooo** [comment]
IMO, as long as the Envoy client can monitor the number of active requests in the LLM runtime, it will work fine.

**Comment by orange-puff** [comment]
Is there a document anywhere that ties load reporting, configuration, and load balancing decisions together? A simpler question is; if I was interested in configuring Envoy to make load balancing decisions on things like upstream reported metrics (e.g. cpu utilization, mem utilization, etc.) where could I start looking into that?

**Comment by efimki** [comment]
> Is there another proposed load balancer this would fit under or is there a desire to consider LocalityLbConfig? I'm interested in contributing towards additional ORCA support but would like to understand the current state and outstanding features.

I'm working on a simple implementation of WrrLocality lb policy that will layer on top of ClientSideWeightedRoundRobin to enable locality weights use in lb decisions.

**Comment by efimki** [comment]
> Is ORCA now widely supported by various LLM Inference runtime?

There is ORCA metrics support in Triton LLM: https://github.com/triton-inference-server/server/pull/8111

I'm working on support for VLLM but it needs more work before merging: https://github.com/vllm-project/vllm/pull/20088

**Comment by efimki** [comment]
> Is there a document anywhere that ties load reporting, configuration, and load balancing decisions together? A simpler question is; if I was interested in configuring Envoy to make load balancing decisions on things like upstream reported metrics (e.g. cpu utilization, mem utilization, etc.) where could I start looking into that?

This is a great question, but unfortunately I'm not aware of such a doc.

**Comment by jukie** [comment]
Thanks, @efimki! I saw https://github.com/envoyproxy/envoy/pull/40577 which adds support for `localityWeightedLbConfig` but my thought here was more about supporting `localityLbConfig` in general, so that users could configure either `localityWeightedLbConfig` or `zoneAwareLbConfig`.

For my use case, I’d like to use `zoneAwareLbConfig` in an environment where inbound traffic is uneven across locality zones. Today, using something like least request + zone-aware routing doesn’t achieve good distribution since Envoy doesn’t coordinate load across proxy instances and assumes inbound traffic is balanced. With ORCA and client-side load balancing, upstream endpoints could report load, allowing for equal global distribution and each proxy instance still prioritizing local traffic.

**Comment by jukie** [comment]
I've been digging into a few ways to address my issue above and I was looking for more input on the below options I could think of:
* Add locality awareness as an option to Client-Side WRR which could be done by either adding direct support for localityLbConfig or design something similar purpose built for ORCA metrics.
* 2 new LbPolicies
  * One that supports localityWeightedLbConfig
  * A second that supports `zoneAwareLbConfig` (to address the uneven imbound traffic would add an ORCA locality_basis mode)
* 1 new LbPolicy that suppots localityLbConfig directly

**Comment by markdroth** [comment]
We should definitely not add locality awareness to `ClientSideWeightedRoundRobin`.  Instead, Envoy should implement the `WrrLocality` LB policy.

Work on this has started in #40577, but I don't know if there's a timeline for that.

**Comment by jukie** [comment]
If `localityLbConfig` could be supported there instead of only `localityWeightedLbConfig`/`commonLbConfig` that would make it the most flexible. I could then take a stab at adding an ORCA locality_basis mode to zoneAwareLbConfig that _should_ slot in nicely to `WrrLocality`.

Edit: now I see that still depends on `ClientSideWeightedRoundRobin` so probably won't work how I was thinking.

**Comment by jukie** [comment]
Hey @efimki, I'm curious if you'd be open to a refactor of wrr_locality to be a standalone lb policy vs a simple wrapper around ClientSideWeightedRoundRobin? The missing item for me is full support for `localityLbConfig` (I want to enable zone aware routing) and the forced `localityWeightedLbConfig` mode. Adding a separate policy for each combination of ORCA metrics and locality routing doesn't scale well, and wrr_locality seems like the natural place to consolidate this. I have a working prototype that's gone through a few iterations but wanted to check if that's the right approach.

**Comment by jukie** [comment]
Have there been any discussions or open plans to implement ORCA out-of-band reporting? The proto field is there for CSWRR but it's not actually implemented.

**Comment by efimki** [comment]
I think implementing out-of-band reporting would be interesting, but currently it is NOT on my radar.

**Comment by altaiezior** [comment]
@efimki @jukie I can help with implementing out-of-band reporting and will use the gRPC library as a reference.

From my understanding, the application exposes metrics via a gRPC service, and Envoy connects to it (likely via a server-streaming call) to receive load reports and adjust weights.

One thing I’m unclear about is how the endpoint for this service should be configured. Since Envoy might serve both gRPC and HTTP for the same endpoint, it seems like this needs to be configured independently. Does that align with your expectations?

**Comment by jukie** [comment]
I've already started working on this in a few draft PRs but will open a separate issue to align on the design.

**Comment by altaiezior** [comment]
Sure let me know @jukie if you need any help with this, lets work together again :D

**Comment by jukie** [comment]
Is there any additional work needed for ORCA or is everything fully implemented?
